In [11]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import scipy.stats as stats

# 
file = pd.read_csv('salary.csv')
file = file.dropna()
file.describe()

salary     exprior  yearsworked   yearsrank      market  \
count    513.000000  513.000000   513.000000  513.000000  513.000000   
mean   50863.220098    2.931774    12.836257    7.046784    0.948558   
std    12685.132358    4.794327     9.447040    6.419708    0.149524   
min    29000.000000    0.000000     0.000000    0.000000    0.710000   
25%    40000.000000    0.000000     4.000000    2.000000    0.840000   
50%    50096.000000    1.000000    12.000000    5.000000    0.920000   
75%    60345.000000    4.000000    22.000000   12.000000    1.020000   
max    96156.000000   25.000000    41.000000   28.000000    1.330000   

           degree   otherqual    position        male       Field    yearsabs  
count  513.000000  513.000000  513.000000  513.000000  513.000000  513.000000  
mean     0.964912    0.044834    2.130604    0.750487    2.526316    6.994152  
std      0.184181    0.207142    0.819977    0.433153    1.126644   16.887552  
min      0.000000    0.000000    1.000000    0.000000    1.000000    0.000000  
25%      1.000000    0.000000    1.000000    1.000000    2.000000    0.000000  
50%      1.000000    0.000000    2.000000    1.000000    3.000000    1.000000  
75%      1.000000    0.000000    3.000000    1.000000    4.000000    2.000000  
max      1.000000    1.000000    3.000000    1.000000    4.000000  118.000000

This is the descriptive stats bla bla bla

## (a) Run a simple linear regression for Salary with one predictor variable: Years Worked

**(i)	What percentage of the variance in employees’ salaries is accounted for by the number of years they have worked?**

What is the p-value? It is not shown in the report

In [8]:
# calculate Pearson's correlation coefficient 
x = file['yearsworked']
y= file['salary']
slope, intercept, r, p, std_err = stats.linregress(x,y)
std = np.std(y) #why is this here - please comment
louise = r**2 * 100 #coefficient of determination
print("percentage variance in salary predicted by variance in years worked = " + str(louise))
print("correlation coefficient = " + str(r))

# What is the output below? what does it mean? How do you interepret it? 
# Write this up as a report, for someone who doesn't know the data or the specific procedures

percentage variance in salary predicted by variance in years worked = 38.88630734995994
correlation coefficient = 0.6235888657598044


## (ii) Does the model significantly predict the dependent variable? Report the amount of variance explained (R2) and significance value (p) to support your answer.

##### - yes it does, because our p-value is well-below our significance threshold of 0.05 . Meaning our coefficient of determination (0.3886307) is most likely to be correct.  

# We can never know the correctness, so rather say that there is a relationship between Years Worked and Salary

## iii)	Report and interpret the standard error of the estimate. (Hint: Compare the standard error of the estimate to the standard deviation for Salary, which you would have calculated in Question 1.a))

In [3]:
print('std :',std)
print('std_err :',std_err)

std : 12672.762650138267
std_err : 46.43634621012888


#####  the avarage difference of the predicted amounts to the actual amount in the "yearsworked" and ''salary'' is 46.44 .

The SEE is an indication of accuracy, as it is the standard deviation of the residuals.  So rather the accuracy of predicting salary by using Years Worked is much better than the accuracy of predicting salary purely based on mean salary. 

It's easy to speak colloquially about average difference, but really it is the the square root of the average squared difference between predicted and observed salary. That's a more accurate way to put it.

Spell check and no bold for normal text please. 

# b)	Examine the output from the Coefficients table:

## (i)	What does the unstandardized coefficient (B) tell you about the relationship between Years Worked and Salary?

In [4]:
sample = pd.DataFrame(file["yearsworked"]).sort_values('yearsworked').reset_index()
sample = sample.loc[[6,44, 50, 100, 130, 150]]
predicted_salary = 0

def predict_salary(yearsworked):
    return slope * year + intercept

for year in sample.yearsworked:
    last_salary = predicted_salary
    predicted_salary = slope * year + intercept
    difference = predicted_salary - last_salary
    print("Years Worked: " + str(year) + " | Predicted Salary = " + str(predicted_salary) + " | Difference = " + str(difference))

Years Worked: 0 | Predicted Salary = 40115.0137875915 | Difference = 40115.0137875915
Years Worked: 1 | Predicted Salary = 40952.34557759235 | Difference = 837.3317900008478
Years Worked: 2 | Predicted Salary = 41789.67736759319 | Difference = 837.3317900008406
Years Worked: 3 | Predicted Salary = 42627.00915759404 | Difference = 837.3317900008478
Years Worked: 4 | Predicted Salary = 43464.34094759488 | Difference = 837.3317900008406
Years Worked: 5 | Predicted Salary = 44301.67273759573 | Difference = 837.3317900008478


##### It tells us that for every ADDITIONAL year worked an employee earns R837 more per annum. 
and what are the b and beta vlaues> it is not shown in the report

## (ii)	Compare the standardised coefficient (Beta) for Years Worked to the Pearson’s (zero-order) correlation coefficient from question 1.c). What do you notice? Explain why this is so. 


In [16]:
std_x = np.std(x)
std_y = np.std(y)

#Beta should be <1 in simple regression

predicted_salary_Std = slope * std_x + intercept
predicted_salary_Std = predicted_salary_Std / std_y
print(predicted_salary_Std) #why is this important? explain
print(r)

#to get the standardised coefficient, you need to standardise x and y .i.e get zscores and then repeat the regression

3.789040227476409
0.6235888657598045


#### We notice that there is a positive relation between X and Y, because our r value (0.62)  shows that there is a positive relationship between our variables(x(years worked) and y(Salary)) and our Standardised coefficient (Beta) shows that when there is a 1 std increase in the years worked there is an increase of 3.75 std in the Salary earned. 
-- our standardised coefficient can be larger than 1, but our correlation coefficient cannot be larger than one. 

## (iii)	Calculate the expected salary for someone who graduated 12 years ago.

In [20]:
Predicted_salary_12 = slope * 12 + intercept
print(Predicted_salary_12 )

50162.99526760164


## (iv)	Calculate the expected salary for someone who graduated 80 years ago. Are there any problems with this prediction? If so, what are they?

In [28]:
p80 = slope * 80 + intercept
print(p80)

107101.55698765905


* No, there is no problem.

#how likely is it that someone who graduated 80 years ago is still alive and working today? They would be over a hundred, at least. Never extrapolate - in other words, don't make guesses based on predictor values outside the range of values you have observed.

# c)	We have only looked at the number of years and employee has worked. What are employee characteristics might influence their salary?


#### Position, Field of work, gender, level of education, experience prior to current position, market value and years worked at current position could all influence salary.

Make this report pretty and easily readable. Assignment title and your name at the top and full sentence explanations of the answers.

Good work on getting the linear regression, but make sure you know how to get the standardised coefficient